src/main/java/uk/gov/justice/digital/writer/structured/StructuredZoneCdcWriter.java

```diff
@@ -49,4 +59,31 @@ public void writeInvalidRecords(
         storage.updateDeltaManifestForTable(spark, destinationPath);
     }
 
+    private static Dataset<Row> getMostRecentUniqueRecords(SparkSession spark, SourceReference.PrimaryKey primaryKey, Dataset<Row> validRecords) {
+        val primaryKeys = JavaConverters
+                .asScalaIteratorConverter(primaryKey.keys.stream().map(functions::col).iterator())
+                .asScala()
+                .toSeq();
+
+        val mostRecentColumnExpressions = Arrays.stream(validRecords.columns())
+                .filter(column -> !primaryKey.keys.contains(column))
+                .map(column -> last(column).as(column));
+
+        val mostRecentColumnExpressionsAsScalaSeq = JavaConverters
+                .asScalaIteratorConverter(mostRecentColumnExpressions.iterator())
+                .asScala()
+                .toSeq();
+
+        if (mostRecentColumnExpressionsAsScalaSeq.nonEmpty()) {
+            return validRecords
+                    .orderBy(TIMESTAMP)
+                    .groupBy(primaryKeys)
```

Comment:
@koladeadewuyi-moj I have seen some conflicting accounts on whether a DataFrame orderBy followed by a groupBy always preserves the order within the groups and under what circumstances the order is preserved. For example, I've seen claims that if the data was not already partitioned by the grouping key then the order may be jumbled up again during the shuffle to produce the groups and therefore you should also include the grouping keys in the orderby. I think I've also seen claims that even that isn't good enough to guarantee the order is preserved in the groups. I think something also said that if the data spills to disk during the grouping it could break the ordering.

https://medium.com/@davisjustin42/group-by-does-not-maintain-order-in-pyspark-use-a-window-function-instead-9178e70497a9

https://stackoverflow.com/questions/39505599/spark-dataframe-does-groupby-after-orderby-maintain-that-order

Is there any likelihood of taking a value for a column here that isn't from the last row?